File: toy-c-compiler-master/src/main/java/toyc/ir/IRPrinter.java
package toyc.ir;

import java.util.StringJoiner;
import toyc.ir.block.BasicBlock;
import toyc.ir.block.Function;
import toyc.ir.block.Module;
import toyc.ir.inst.Alloca;
import toyc.ir.inst.BinaryOp;
import toyc.ir.inst.Branch;
import toyc.ir.inst.Call;
import toyc.ir.inst.Compare;
import toyc.ir.inst.CondBranch;
import toyc.ir.inst.GlobalAddr;
import toyc.ir.inst.Load;
import toyc.ir.inst.LoadImm;
import toyc.ir.inst.Move;
import toyc.ir.inst.Phi;
import toyc.ir.inst.Return;
import toyc.ir.inst.Store;
import toyc.ir.inst.UnaryOp;
import toyc.ir.value.Constant;
import toyc.ir.value.GlobalVar;
import toyc.ir.value.Label;
import toyc.ir.value.LocalVar;
import toyc.ir.value.Temp;

public final class IRPrinter implements IRVisitor<Void, Integer> {
    private static final String INDENT = "  ";
    private final StringBuilder sb = new StringBuilder();

    private IRPrinter() {
    }

    public static String print(IRProgram program) {
        IRPrinter printer = new IRPrinter();
        program.accept(printer, 0);
        return printer.sb.toString();
    }

    private void emit(int depth, String text) {
        sb.append(INDENT.repeat(depth)).append(text).append('\n');
    }

    @Override
    public Void visitProgram(IRProgram node, Integer depth) {
        emit(depth, "IRProgram");
        node.module().accept(this, depth + 1);
        return null;
    }

    @Override
    public Void visitModule(Module node, Integer depth) {
        emit(depth, "Module");
        for (GlobalVar global : node.globals()) {
            global.accept(this, depth + 1);
        }
        for (Constant constant : node.globalConstants()) {
            constant.accept(this, depth + 1);
        }
        for (Function function : node.functions()) {
            function.accept(this, depth + 1);
        }
        return null;
    }

    @Override
    public Void visitFunction(Function node, Integer depth) {
        StringJoiner params = new StringJoiner(", ");
        for (LocalVar parameter : node.parameters()) {
            params.add(parameter.name());
        }
        emit(depth, "Function " + node.returnType() + " " + node.name() + "(" + params + ")");
        for (BasicBlock block : node.blocks()) {
            block.accept(this, depth + 1);
        }
        return null;
    }

    @Override
    public Void visitBasicBlock(BasicBlock node, Integer depth) {
        emit(depth, "Block " + node.label().name());
        node.allInstructions().forEach(instruction -> instruction.accept(this, depth + 1));
        return null;
    }

    @Override
    public Void visitConstant(Constant node, Integer depth) {
        emit(depth, "Constant " + node.name() + " = " + node.value());
        return null;
    }

    @Override
    public Void visitGlobalVar(GlobalVar node, Integer depth) {
        emit(depth, "GlobalVar " + node.name() + " = " + node.initialValue());
        return null;
    }

    @Override
    public Void visitLabel(Label node, Integer depth) {
        emit(depth, "Label " + node.name());
        return null;
    }

    @Override
    public Void visitLocalVar(LocalVar node, Integer depth) {
        emit(depth, "LocalVar " + node.name());
        return null;
    }

    @Override
    public Void visitTemp(Temp node, Integer depth) {
        emit(depth, "Temp " + node.name());
        return null;
    }

    @Override
    public Void visitAlloca(Alloca node, Integer depth) {
        emit(depth, node.result().name() + " = alloca");
        return null;
    }

    @Override
    public Void visitBinaryOp(BinaryOp node, Integer depth) {
        emit(depth, node.result().name() + " = " + node.op() + " " + node.left() + ", " + node.right());
        return null;
    }

    @Override
    public Void visitCompare(Compare node, Integer depth) {
        emit(depth, node.result().name() + " = cmp " + node.predicate() + " " + node.left() + ", " + node.right());
        return null;
    }

    @Override
    public Void visitUnaryOp(UnaryOp node, Integer depth) {
        emit(depth, node.result().name() + " = " + node.op() + " " + node.value());
        return null;
    }

    @Override
    public Void visitLoadImm(LoadImm node, Integer depth) {
        emit(depth, node.result().name() + " = imm " + node.constant().value());
        return null;
    }

    @Override
    public Void visitMove(Move node, Integer depth) {
        emit(depth, node.result().name() + " = move " + node.value());
        return null;
    }

    @Override
    public Void visitPhi(Phi node, Integer depth) {
        StringJoiner incoming = new StringJoiner(", ");
        for (Phi.Incoming value : node.incoming()) {
            incoming.add("[" + value.value() + ", " + value.predecessor().name() + "]");
        }
        emit(depth, node.result().name() + " = phi " + incoming);
        return null;
    }

    @Override
    public Void visitLoad(Load node, Integer depth) {
        emit(depth, node.result().name() + " = load " + node.address());
        return null;
    }

    @Override
    public Void visitStore(Store node, Integer depth) {
        emit(depth, "store " + node.value() + ", " + node.address());
        return null;
    }

    @Override
    public Void visitGlobalAddr(GlobalAddr node, Integer depth) {
        emit(depth, node.result().name() + " = globaladdr " + node.global().name());
        return null;
    }

    @Override
    public Void visitCall(Call node, Integer depth) {
        StringJoiner args = new StringJoiner(", ");
        for (var arg : node.args()) {
            args.add(arg.name());
        }
        String prefix = node.result() == null ? "" : node.result().name() + " = ";
        emit(depth, prefix + "call " + node.returnType() + " " + node.functionName() + "(" + args + ")");
        return null;
    }

    @Override
    public Void visitBranch(Branch node, Integer depth) {
        emit(depth, "br " + node.target().name());
        return null;
    }

    @Override
    public Void visitCondBranch(CondBranch node, Integer depth) {
        emit(depth, "cbr " + node.condition() + ", " + node.trueTarget().name() + ", " + node.falseTarget().name());
        return null;
    }

    @Override
    public Void visitReturn(Return node, Integer depth) {
        emit(depth, node.value() == null ? "ret" : "ret " + node.value());
        return null;
    }
}

File: src/ir/ir_printer.cpp
#include "toyc/ir/ir_printer.h"

#include <sstream>
#include <string>

// ===========================================================================
// Helpers
// ===========================================================================

static void indent(std::ostream& os, int depth) {
    for (int i = 0; i < depth; ++i) os << "  ";
}

template <typename Container, typename Accessor>
static void print_seq(std::ostream& os, const Container& c, Accessor&& acc,
                      const char* sep = ", ") {
    bool first = true;
    for (const auto& elem : c) {
        if (!first) os << sep;
        os << acc(elem);
        first = false;
    }
}

// ===========================================================================
// IRPrinter::print(Value)
// ===========================================================================

std::string IRPrinter::print(const Value& value) {
    return value.name();
}

// ===========================================================================
// IRPrinter::print(Instr)
// ===========================================================================

std::string IRPrinter::print(const Instr& instr) {
    std::ostringstream os;

    switch (instr.kind()) {
    case InstrKind::Alloca: {
        auto& a = static_cast<const AllocaInstr&>(instr);
        os << a.local()->name() << " = alloca";
        break;
    }
    case InstrKind::Load: {
        auto& l = static_cast<const LoadInstr&>(instr);
        os << l.temp()->name() << " = load " << l.address()->name();
        break;
    }
    case InstrKind::Store: {
        auto& s = static_cast<const StoreInstr&>(instr);
        os << "store " << s.value()->name() << ", " << s.address()->name();
        break;
    }
    case InstrKind::LoadImm: {
        auto& l = static_cast<const LoadImmInstr&>(instr);
        os << l.temp()->name() << " = imm " << l.int_value();
        break;
    }
    case InstrKind::BinaryOp: {
        auto& b = static_cast<const BinaryOpInstr&>(instr);
        os << b.temp()->name() << " = " << BinaryOpInstr::op_name(b.op())
           << " " << b.left()->name() << ", " << b.right()->name();
        break;
    }
    case InstrKind::UnaryOp: {
        auto& u = static_cast<const UnaryOpInstr&>(instr);
        os << u.temp()->name() << " = " << UnaryOpInstr::op_name(u.op())
           << " " << u.value()->name();
        break;
    }
    case InstrKind::Compare: {
        auto& c = static_cast<const CompareInstr&>(instr);
        os << c.temp()->name() << " = cmp " << CompareInstr::pred_name(c.predicate())
           << " " << c.left()->name() << ", " << c.right()->name();
        break;
    }
    case InstrKind::Call: {
        auto& c = static_cast<const CallInstr&>(instr);
        if (c.result()) {
            os << c.temp()->name() << " = ";
        }
        os << "call " << to_string(c.return_type()) << " " << c.callee() << "(";
        print_seq(os, c.args(), [](Value* v) { return v->name(); });
        os << ")";
        break;
    }
    case InstrKind::Move: {
        auto& m = static_cast<const MoveInstr&>(instr);
        os << m.temp()->name() << " = move " << m.value()->name();
        break;
    }
    case InstrKind::Phi: {
        auto& p = static_cast<const PhiInstr&>(instr);
        os << p.temp()->name() << " = phi";
        bool first = true;
        for (const auto& inc : p.incoming()) {
            if (!first) os << ",";
            os << " [" << inc.value->name() << ", " << inc.predecessor->name() << "]";
            first = false;
        }
        break;
    }
    case InstrKind::Branch: {
        auto& b = static_cast<const BranchInstr&>(instr);
        os << "br " << b.target()->name();
        break;
    }
    case InstrKind::CondBranch: {
        auto& c = static_cast<const CondBranchInstr&>(instr);
        os << "cbr " << c.condition()->name() << ", "
           << c.true_target()->name() << ", " << c.false_target()->name();
        break;
    }
    case InstrKind::Return: {
        auto& r = static_cast<const ReturnInstr&>(instr);
        if (r.has_value()) {
            os << "ret " << r.value()->name();
        } else {
            os << "ret";
        }
        break;
    }
    case InstrKind::GlobalAddr: {
        auto& g = static_cast<const GlobalAddrInstr&>(instr);
        os << g.temp()->name() << " = globaladdr " << g.global()->name();
        break;
    }
    }

    return os.str();
}

// ===========================================================================
// IRPrinter::print(BasicBlock)
// ===========================================================================

std::string IRPrinter::print(const BasicBlock& block) {
    std::ostringstream os;
    os << block.label()->name() << ":\n";

    for (auto* instr : block.all_instrs()) {
        os << "  " << IRPrinter::print(*instr) << "\n";
    }

    return os.str();
}

// ===========================================================================
// IRPrinter::print(Function)
// ===========================================================================

std::string IRPrinter::print(const Function& function) {
    std::ostringstream os;

    // Function header.
    os << "func @" << function.name() << "() -> " << to_string(function.return_type());
    os << " {\n";

    // Parameters / locals declaration.
    for (auto& [name, local] : function.locals()) {
        (void)name;
        if (!local->is_parameter()) {
            os << "  local " << local->name() << "\n";
        }
    }

    // Blocks in order.
    for (const auto& block : function.blocks()) {
        os << IRPrinter::print(*block);
    }

    os << "}\n";
    return os.str();
}

// ===========================================================================
// IRPrinter::print(Module)
// ===========================================================================

std::string IRPrinter::print(const Module& module) {
    std::ostringstream os;
    os << "module {\n";

    // Globals
    for (auto* gv : module.globals()) {
        os << "  global " << gv->name() << " = " << gv->initial_value() << "\n";
    }

    // Global constants
    for (auto* c : module.global_constants()) {
        os << "  const " << c->name() << " = " << c->value() << "\n";
    }

    os << "}\n";

    // Functions
    for (const auto& fn : module.functions()) {
        os << "\n" << IRPrinter::print(*fn);
    }

    return os.str();
}

// ===========================================================================
// IRPrinter::print(IRProgram)
// ===========================================================================

std::string IRPrinter::print(const IRProgram& program) {
    return IRPrinter::print(*program.module());
}
